src/economy.cpp

```diff
@@ -103,14 +103,38 @@ static PriceMultipliers _price_base_multiplier;
 
 /**
  * Calculate the value of the company. That is the value of all
- * assets (vehicles, stations, etc) and money minus the loan,
+ * assets (vehicles, stations, shares) and money minus the loan,
  * except when including_loan is \c false which is useful when
  * we want to calculate the value for bankruptcy.
- * @param c              the company to get the value of.
+ * @param c the company to get the value of.
  * @param including_loan include the loan in the company value.
  * @return the value of the company.
  */
-Money CalculateCompanyValue(const Company *c, bool including_loan)
+Money CalculateCompanyValue(const Company* c, bool including_loan)
+{
+	Money owned_shares_value = 0;
+
+	uint8 shares_owned = 0;
+
+	for (const Company *co : Company::Iterate()) {
+
+		shares_owned = 0;
+
+		for (uint8 i = 0; i < 4; i++) {
+			if (co->share_owners[i] == c->index) {
+				shares_owned++;
+			}
+		}
+
+		if (shares_owned > 0) {
+			owned_shares_value += (CalculateCompanyValueExcludingShares(co) / 4) * shares_owned;
+		}
```

Comment:
probably not worth bothering with the if statement here - CalculateCompanyValueExcludingShares isn't a particularly costly function to call needlessly (if `shares_owned == 0`)

Reply:
CalculateCompanyValueExcludingShares iterates all vehicles and stations so it is somewhat costly. Enough to add a check to skip it in the most common case.